### DISCUSSION TITLE
Different Router Types for external and internal usage in a micro service architecture

### DISCUSSION THREAD

**Discussion by cimchd** [trpc/trpc#5258] [Q&A]
We have a micro service architecture inside a monorepo and all our services are using trpc. Some services can be called by other internal micro services as well as some external services.
If called from an external service we want to hide some fields. Here is a simplified code example to get the idea:
```ts
// router.ts
// [...]
export const appRouter = router({
  getAttributes: publicProcedure
    .input( /* [...] */)
    .query(async ({ ctx, input }) => {
      if (ctx.internalRequest) {
        return {
          publicAttribute: 1,
          internalAttribute: 2
        }
      } else {
        return {
          publicAttribute: 1
        }
      }
    }),
// [...]
});
```
Our problem is that the type of the route is not conditional (in our case it's always the type for the external route with less attributes). As far as we know typescript does not support conditional types at the moment (see: https://github.com/trpc/trpc/discussions/2888).

It seems that we have the following options:

**1. Create two queries (one for internal and one for external)**
   This has the drawback that we need to duplicate code. Especially for complex queries this is something we want to avoid.

**2. Typecast the correct type manually on one side**
   This has the drawback that we have to typecast every usage of the query (besides that a typecast is only a workaround)

**3. Create two router types**
   This is a variation of 2. but instead of typecasting every query usage we could change the Router type for on side:
```ts
import { appRouter } from './trpc/router.js';

export type InternalRouter = typeof appRouter;
export type ExternalRouter = CreateExternalRouter<typeof appRouter>;
```
This has the advantage that we only need to fix the type once and the usage of the query stays the same as long as they import the correct type.
The type `CreateExternalRouter` is basically a DeepOmit type that omits the attributes we don't want to have inside the external router.
Our problem here is that we get an error message: 

```
Property 'getAttributes' does not exist on type '"The property 'useContext' in your router collides with a built-in method, rename this router or procedure on your backend." | "The property 'useUtils' in your router collides with a built-in method, rename this router or procedure on your backend." | "The property 'Provider' in your router collides with a built-in ...'.
  Property 'customer' does not exist on type '"The property 'useContext' in your router collides with a built-in method, rename this router or procedure on your backend."'.ts(2339)
```
It seems that our External Router type is not compatible with the trpc client.

So our questions are:
1. Is there a better option than the above 3?
2. Or is there a way to change the Router type in a way that is compatible with the trpc client?

I tried to find a solution for this for several days now. Any help or suggestions are welcome :-)

**Comment by sebasptsch** [comment]
You can create a Zod union of the two response schemas and then create a check that eliminates one of the types.

**Reply by cimchd** [reply]
Thanks for your reply. I don't think this will help us in our case. We will have the same problem with the resulting typescript type of the query procedure.

**Reply by sebasptsch** [reply]
Hmm, making another router and abstracting the function it calls out to a seperate file may be the way to go then.
